s/core.go

```diff
@@ -0,0 +1,269 @@
+	return sup.spec.Name()
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Spec (static tree) functionality
+
+// emptyEventNotifier is an utility function that works as a default value
+// whenever an EventNotifier is not specified on the Supervisor Spec
+func emptyEventNotifier(_ Event) {}
+
+// getEventNotifier returns the configured EventNotifier or emptyEventNotifier
+// (if none is given via WithEventNotifier)
+func (spec Spec) getEventNotifier() EventNotifier {
+	if spec.eventNotifier == nil {
+		return emptyEventNotifier
+	}
+	return spec.eventNotifier
+}
+
+// subtreeMain contains the main logic of the Child spec that runs a supervision
+// sub-tree. It returns an error if the child supervisor fails to start.
+func subtreeMain(parentName string, spec Spec) func(context.Context, func()) error {
+	// we use the start version that receives the notifyChildStart callback, this
+	// is essential, as we need this callback to signal the sub-tree children have
+	// started before signaling we have started
+	return func(parentCtx context.Context, notifyChildStart func()) error {
+		// in this function we use the private versions of start and wait
+		// given we don't want to signal the eventNotifier more than once
+		// on sub-trees
+
+		ctx, cancelFn := context.WithCancel(parentCtx)
+		defer cancelFn()
+		sup, err := spec.start(ctx, parentName)
+		if err != nil {
+			return err
+		}
+		notifyChildStart()
+		return sup.wait()
+	}
+}
+
+// Subtree allows to register a Supervisor Spec as a sub-tree of a bigger
+// Supervisor Spec.
+func (spec Spec) Subtree(subtreeSpec Spec, copts ...c.Opt) c.Spec {
+	subtreeSpec.eventNotifier = spec.eventNotifier
+	return c.New1(subtreeSpec.Name(), subtreeMain(spec.name, subtreeSpec), copts...)
+}
+
+// start is routine that contains the main logic of a Supervisor. This function:
+//
+// 1) spawns a new goroutine for the supervisor loop
+//
+// 2) spawns each child goroutine in the correct order
+//
+// 3) stops all the spawned children in the correct order once it gets a stop
+// signal
+//
+// 4) it monitors and reacts to errors reported by the supervised children
+//
+func (spec Spec) start(parentCtx context.Context, parentName string) (Supervisor, error) {
+	// cancelFn is used when Stop is requested
+	ctx, cancelFn := context.WithCancel(parentCtx)
+
+	// evCh is used to keep track of errors from children
+	// evCh := make(chan ChildEvent)
+
+	// ctrlCh is used to keep track of request from client APIs (e.g. spawn child)
+	// ctrlCh := make(chan ControlMsg)
+
+	// startCh is used to track when the supervisor loop thread has started
+	startCh := make(chan struct{})
+
+	// terminateCh is used when waiting for cancelFn to complete
+	terminateCh := make(chan struct{})
+
+	// errCh is used when supervisor gives up on error handling
+	errCh := make(chan error)
+
+	eventNotifier := spec.getEventNotifier()
+
+	var runtimeName string
+	if parentName == "" {
```

Comment:
I think this could be made declarative with a `RootSpec` type.

Reply:
Adding the `RootSpec` type adds a bunch of other complications, I added a constant to make clearer in the code what is going on, and leaving it as is for now